File: flink16-examples/src/main/scala/com/flink/pipeline/example4/FlinkPipelineExample1.scala
package com.flink.pipeline.example4

import com.turing.common.FlinkEnvUtils
import com.turing.pipeline.example3.FlinkPipelineExample
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend
import org.slf4j.LoggerFactory

object FlinkPipelineExample1 {

  private val logger = LoggerFactory.getLogger(this.getClass.getName)

  def main(args: Array[String]): Unit = {

    val fileName = "application-dev.properties"
    val inputStream = classOf[FlinkPipelineExample].getClassLoader.getResourceAsStream(fileName)
    val parameterTool = ParameterTool.fromPropertiesFile(inputStream)
    val parallelisNum = Integer.valueOf(parameterTool.get("flink.pipeline.parallelism", "1"))
    val checkpointPath = parameterTool.get("flink.pipeline.checkpoint.url")
    logger.info("load global property file:{}", fileName)

    val flinkEnv = FlinkEnvUtils.getStreamTableEnv(args)
    flinkEnv.env.setParallelism(parallelisNum)
    flinkEnv.env.setStateBackend(new EmbeddedRocksDBStateBackend)
    flinkEnv.env.getCheckpointConfig.setCheckpointStorage(checkpointPath)

   val executeSql =
     """
       |CREATE TEMPORARY TABLE location_updates (
       |  `character_id` INT,
       |  `location` STRING,
       |  `proctime` AS PROCTIME()
       |)
       |WITH (
       |  'connector' = 'faker',
       |  'fields.character_id.expression' = '#{number.numberBetween ''0'',''100''}',
       |  'fields.location.expression' = '#{harry_potter.location}'
       |);
       |CREATE TEMPORARY TABLE characters (
       |  `character_id` INT,
       |  `name` STRING
       |)
       |WITH (
       |  'connector' = 'faker',
       |  'fields.character_id.expression' = '#{number.numberBetween ''0'',''100''}',
       |  'fields.name.expression' = '#{harry_potter.characters}'
       |);
       |CREATE TABLE sink_table (
       |    character_id INT,
       |    location STRING,
       |    name STRING
       |) WITH (
       |  'connector' = 'print'
       |);
       |INSERT INTO sink_table
       |SELECT
       |  c.character_id,
       |  l.location,
       |  c.name
       |FROM location_updates AS l
       |JOIN characters FOR SYSTEM_TIME AS OF proctime AS c
       |ON l.character_id = c.character_id;
       |""".stripMargin

    val executeSqlArrary = executeSql.split(";")

    /**
     * 算子 {@link org.apache.flink.streaming.api.operators.KeyedProcessOperator}
     * --  {@link org.apache.flink.table.runtime.operators.deduplicate.ProcTimeDeduplicateKeepFirstRowFunction}
     */
    for (innerSql <- executeSqlArrary) {
        flinkEnv.streamTEnv.executeSql(innerSql)
    }

    flinkEnv.env.execute("FlinkPipelineExample")

  }

}

File: flink16-examples/src/main/java/com/turing/pipeline/example5/FlinkPipelineExample1.java
package com.turing.pipeline.example5;

import com.turing.common.FlinkEnvUtils;
import com.turing.pipeline.example4.KafkaUtilsExample4;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend;
import org.apache.flink.table.api.Table;
import org.apache.flink.types.Row;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;

/**
 * @descri
 *
 * @author lj.michale
 * @date 2023-03-10
 */
public class FlinkPipelineExample1 {

    private static final Logger logger = LoggerFactory.getLogger(KafkaUtilsExample4.class);

    public static void main(String[] args) throws Exception {

        final String fileName = "application-dev.properties";
        InputStream inputStream = com.turing.pipeline.example3.FlinkPipelineExample.class.getClassLoader().getResourceAsStream(fileName);
        ParameterTool parameterTool = ParameterTool.fromPropertiesFile(inputStream);
        int parallelisNum = Integer.valueOf(parameterTool.get("flink.pipeline.parallelism","1"));
        String checkpointPath = parameterTool.get("flink.pipeline.checkpoint.url");
        logger.info("flink.pipeline.parallelism:{}", parallelisNum);

        FlinkEnvUtils.FlinkEnv flinkEnv = FlinkEnvUtils.getStreamTableEnv(args);
        flinkEnv.env().setParallelism(parallelisNum);
        // 状态后端使用RocksDB
        flinkEnv.env().setStateBackend(new EmbeddedRocksDBStateBackend());
        flinkEnv.env().getCheckpointConfig().setCheckpointStorage(checkpointPath);

        String createTableSql = "CREATE TABLE spells_cast (\n" +
                "    wizard STRING,\n" +
                "    spell  STRING\n" +
                ") WITH (\n" +
                "  'connector' = 'faker',\n" +
                "  'fields.wizard.expression' = '#{harry_potter.characters}',\n" +
                "  'fields.spell.expression' = '#{harry_potter.spells}'\n" +
                ");\n";

        String querySql = "SELECT wizard, spell, times_cast\n" +
                "FROM (\n" +
                "    SELECT *,\n" +
                "    ROW_NUMBER() OVER (PARTITION BY wizard ORDER BY times_cast DESC) AS row_num\n" +
                "    FROM (SELECT wizard, spell, COUNT(*) AS times_cast FROM spells_cast GROUP BY wizard, spell)\n" +
                ")\n" +
                "WHERE row_num <= 2;\n";

        flinkEnv.streamTEnv().executeSql(createTableSql);
        Table resultTable = flinkEnv.streamTEnv().sqlQuery(querySql);

        flinkEnv.streamTEnv()
                .toDataStream(resultTable, Row.class).print();

        flinkEnv.env().execute("KafkaUtilsExample4");
    }
}
